reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,263 @@
+
+Let's take `bytearray` as an example.
+
+```python
+# In Python
+class ByteArraySerializer(RaySerializer):
+
+    def serialize(self, byte_array: bytearray) -> RaySerializationResult:
+        return RaySerializationResult(None, [memoryview(byte_array)])
+
+    def deserialize(self, serialization_result: RaySerializationResult,
+                    oob_offset: int = 0) -> Tuple[bytearray, int]:
+        return (serialization_result.out_of_band_buffers[0].obj, oob_offset)
+```
+
+Now bytes object will be out-of-band serialized.
+
+In the same process, no copy will happen to out-of-band buffers.
+
+For example, in the following code, no copy will happen:
+
+```python
+b = bytearray(b'\x01\x02\x03')
+serialization_result: RaySerializationResult = ray.serialize(b)
+deserialized_bytes: bytearray = ray.deserialize(serialization_result)
+```
+
+`deserialized_bytes` and `b` will be backed by the same memory.
+
+When the serialized result needs to go through network, for example, in cross-language situation, only 1 copy will happen. That is from `RaySerializationResult` to the network buffer.
+
+### Nested Serialization
+
+Let's take `List[bytearray]` as an example to see how to handle nested types. We've already implemented a serializer for `bytearray` above.
+
+Note that in real world we might simply wrap Msgpack to do serialization for common classes. Here we rebuild the wheel only to show how it works.
+
+```python
+class ListSerializer(RaySerializer):
+
+    def serialize(self, list_object: List[bytearray]) -> RaySerializationResult:
+        result = RaySerializationResult()
+        result.in_band_buffer = bytearray()
+        # Total element count.
+        result.in_band_buffer.extend(to_bytes(len(list_object))))
+        for element in list_object:
+            element_res = ray.serialize(element)
+            # in band buffer length
+            result.in_band_buffer.extend(to_bytes(len(element_res.in_band_buffer)))
+            # in band buffer
+            result.in_band_buffer.extend(element_res.in_band_buffer)
+            # append out of band buffer
+            result.out_of_band_buffers += element_res.out_of_band_buffers
+        return result
+
+    def deserialize(self, serialization_result: RaySerializationResult,
+                    oob_offset: int = 0) -> Tuple[List[bytearray], int]:
+        in_band = serialization_result.in_band_buffer
+        oob = serialization_result.out_of_band_buffers
+
+        result = []
+        pos = 0
+        # read element count
+        count, pos = get_int(in_band, pos)
+        for i in range(count):
+            # read element size
+            element_size, pos = get_int(in_band, pos)
+            # read element in-band data
+            element_in_band_data = in_band_buffer[pos: pos + element_size]
+            pos = pos + element_size
+            # nested serialization here
+            element, oob_offset = ray.deserialize(element_in_band_data, oob, oob_offset)
```

Comment:
I'm a bit confused what oob_offset is used for in this example. It never seems to be used for deserialization at all, just passed through and stays as 0.

Can you modify the example so it actually demonstrates why you need oob_offset?